drivers/lis3mdl/lis3mdl.c

```diff
@@ -0,0 +1,216 @@
+ *
+ * @author      René Herthel <rene-herthel@outlook.de>
+ *
+ * @}
+ */
+
+#include "lis3mdl.h"
+#include "include/lis3mdl-internal.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+int lis3mdl_init(lis3mdl_t *dev, i2c_t i2c, gpio_t mag_pin, uint8_t mag_address,
+                 lis3mdl_xy_mode_t mag_xy_mode, lis3mdl_z_mode_t mag_z_mode,
+                 lis3mdl_odr_t mag_odr, lis3mdl_scale_t mag_scale)
+{
+    int res;
+    char tmp;
+
+    dev->i2c = i2c;
+    dev->mag_address = mag_address;
+    dev->mag_pin = mag_pin;
+
+    i2c_acquire(dev->i2c);
+    i2c_init_slave(i2c, mag_address);
+    DEBUG("LIS3MDL: Initialize device ... ")
+
+    tmp = (LIS3MDL_MASK_REG2_REBOOT     /* reboot memory content    */
+          | LIS3MDL_MASK_REG2_SOFT_RST  /* reset registers          */
+          | mag_scale);                 /* full-scale configuration */
+
+    res = i2c_write_reg(dev->i2c, dev->mag_address,
+                        LIS3MDL_CTRL_REG2, tmp);
+
+    tmp = (LIS3MDL_MASK_REG1_TEMP_EN  /* enable temperature sensor  */
+          | mag_xy_mode               /* configure x-, y-modes      */
+          | mag_odr);                 /* configure output data rate */
+
+    res += i2c_write_reg(dev->i2c, dev->mag_address,
+                         LIS3MDL_CTRL_REG1, tmp);
+
+    /* configure z-mode */
+    res += i2c_write_reg(dev->i2c, dev->mag_address,
+                         LIS3MDL_CTRL_REG4, mag_z_mode);
+
+    i2c_release(dev->i2c);
+
+    if (res != 3)
+    {
+        DEBUG("[ FAIL ]\n");
+        return -1;
+    }
+
+    DEBUG("[ OK ]\n");
+    return 0;
+}
+
+int lis3mdl_read_mag(lis3mdl_t *dev, lis3mdl_3d_data_t *data)
+{
+    int res;
+    char tmp;
+
+    i2c_acquire(dev->i2c);
+    DEBUG("LIS3MDL: Retrieving magnetometer data ... ");
+
+    /* enable block data update */
+    res = i2c_write_reg(dev->i2c, dev->mag_address,
+                         LIS3MDL_CTRL_REG5, LIS3MDL_MASK_REG5_BDU);
+
+    /* retrieve data from x-axis */
+    res += i2c_read_reg(dev->i2c, dev->mag_address,
+                        LIS3MDL_OUT_X_L_REG, &tmp);
+    data->x_axis = tmp;
+    res += i2c_read_reg(dev->i2c, dev->mag_address,
+                        LIS3MDL_OUT_X_H_REG, &tmp);
+    data->x_axis = tmp<<8; /* shift the 8 msb to left */
+
+    /* retrieve data from y-axis */
+    res += i2c_read_reg(dev->i2c, dev->mag_address,
+                        LIS3MDL_OUT_Y_L_REG, &tmp);
+    data->y_axis = tmp;
+    res += i2c_read_reg(dev->i2c, dev->mag_address,
+                        LIS3MDL_OUT_Y_H_REG, &tmp);
+    data->y_axis = tmp<<8;
+
+    /* retrieve data from z-axis */
+    res += i2c_read_reg(dev->i2c, dev->mag_address,
+                        LIS3MDL_OUT_Z_L_REG, &tmp);
+    data->z_axis = tmp;
+    res += i2c_read_reg(dev->i2c, dev->mag_address,
+                        LIS3MDL_OUT_Z_H_REG, &tmp);
+    data->z_axis = tmp<<8;
+
+    /* disable block data update */
+    res += i2c_write_reg(dev->i2c, dev->mag_address,
```

Comment:
@PeterKietzmann Hmm, okey. Yes two good Points, maybe I should remove this for the Moment, and later, when we can test the device, we can look, if this is a good idea or not.